src/app/auth/Login.tsx

```diff
@@ -0,0 +1,156 @@
+import {
+  faGitlab,
+  faGoogle,
+  faOpenid
+} from '@fortawesome/free-brands-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { toLower, upperFirst } from 'lodash';
+import { useCallback, useEffect, useState } from 'react';
+import { Navigate } from 'react-router-dom';
+import logoFlag from '~/assets/logo-flag.png';
+import { listAuthMethods } from '~/data/api';
+import { useError } from '~/data/hooks/error';
+import { useSession } from '~/data/hooks/session';
+import { AuthMethod, AuthMethodOIDC } from '~/types/Auth';
+
+interface ILoginProvider {
+  displayName: string;
+  icon?: any;
+}
+
+const knownProviders: Record<string, ILoginProvider> = {
+  google: {
+    displayName: 'Google',
+    icon: faGoogle
+  },
+  gitlab: {
+    displayName: 'GitLab',
+    icon: faGitlab
+  },
+  auth0: {
+    displayName: 'Auth0'
+  }
+};
+
+export default function Login() {
+  const { session } = useSession();
+
+  const [providers, setProviders] = useState<
+    {
+      name: string;
+      authorize_url: string;
+      callback_url: string;
+      icon: any;
+    }[]
+  >([]);
+
+  const { setError, clearError } = useError();
+
+  const authorize = async (uri: string) => {
+    const res = await fetch(uri, {
+      credentials: 'include',
+      method: 'GET',
+      headers: {
+        'Content-Type': 'application/json'
+      }
+    });
+
+    if (!res.ok || res.status !== 200) {
+      setError(new Error('Unable to authenticate: ' + res.text()));
+      return;
+    }
+
+    clearError();
+    const body = await res.json();
+    window.location.href = body.authorizeUrl;
+  };
+
+  const loadAvailableProviders = useCallback(async () => {
+    try {
+      const resp = await listAuthMethods();
+      // TODO: support alternative auth methods
+      const authOIDC = resp.methods.find(
+        (m: AuthMethod) => m.method === 'METHOD_OIDC' && m.enabled
+      ) as AuthMethodOIDC;
+
+      if (!authOIDC) {
+        return;
+      }
+
+      const loginProviders = Object.entries(authOIDC.metadata.providers).map(
+        ([k, v]) => {
+          k = toLower(k);
+          return {
+            name: knownProviders[k]?.displayName || upperFirst(k), // if we dont know the provider, just capitalize the first letter
+            authorize_url: v.authorize_url,
+            callback_url: v.callback_url,
+            icon: knownProviders[k]?.icon || faOpenid // if we dont know the provider icon, use the openid icon
+          };
+        }
+      );
+      setProviders(loginProviders);
+    } catch (err) {
+      setError(err instanceof Error ? err : Error(String(err)));
+    }
+  }, [setError]);
+
+  useEffect(() => {
+    loadAvailableProviders();
```

Comment:
This seems like it would cause an endless loop of loading providers?

Reply:
i checked with console.log it only calls once, I'm wondering if its because its wrapped in a `useCallback`?

Reply:
Could this be because you're passing the `loadAvailableProviders` function as an arg below?

`useEffect` docs suggest the second argument is compared between each render.
If it doesn't change then the first argument function is not called again.
Seems to me that function will never change. Or maybe there is some overloaded alternative behaviour when a function is passed as second arg.

Seems you could also pass `[]` for the same effect:
https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects

Reply:
It seems like more often than not you want state in your dependencies array passed to `useEffect`.
There are cases where you might pass a function. Though that seems to be when they're defined out of scope:
https://reacttraining.com/blog/when-to-use-functions-in-hooks-dependency-array/

(I admit as I kept reading this I started to glaze over and maybe missed something key)